mixin/mixout all topologies support

serhiy-katsyuba-intel opened the pull request:
Adds possibility to connect single mixin to multiple mixouts. This allows for all possible mixin/mixout topologies.

RanderWang reviewed (approved):
LGTM

lgirdwood commented on src/audio/mixin_mixout.c:
```diff
@@ -0,0 +1,999 @@
+	source_avail_bytes = audio_stream_get_avail_bytes(&source_c->stream);
+	sinks_free_bytes = INT32_MAX;
+
+	list_for_item(blist, &dev->bsink_list) {
```
Can we comment what this block is doing.

lgirdwood commented on src/audio/mixin_mixout.c:
```diff
@@ -0,0 +1,999 @@
+		struct comp_buffer __sparse_cache *sink_c;
+		uint32_t stream_free_bytes;
+
+		// unused buffer between mixin and mixout
```
C style

lgirdwood commented on src/audio/mixin_mixout.c:
```diff
@@ -0,0 +1,999 @@
+		sink_c = buffer_acquire(sink);
+
+		stream_free_bytes = audio_stream_get_free_bytes(&sink_c->stream);
+		assert(stream_free_bytes >= src_info->consumed_yet_not_produced_bytes);
```
Can this check be non fatal i.e. we should log() the error and return the error up the stack and allow other unrelated pipelines to continue.

serhiy-katsyuba-intel replied:
Actually that condition (stream_free_bytes >= src_info->consumed_yet_not_produced_bytes) is always true. The only possibility it will evaluate to false if someone modified the code and introduce a bug.

So I think it is better to either remove that assert() altogether or leave it as is. But replacing it with if (...) { log_err(...) } seems not a good idea -- the error would just indicate that someone introduced a bug in code and so it is better to signal it in a clear way instead of trying to run other pipelines possibly hiding the issue.

I chose assert() as in case of a bug the math will broke and uint32_t overflow will occur possibly resulting in writing beyond buffer allocated memory which might result in magic issues hard to find.

This assert() is inside the loop, however, this loop iterates by mixouts connected to single mixing. Most often it is just 1 with max 3 (for Meteorlake). So it seems like not much and that assert() can be left as is.

**_What do you think: should I left that assert() as is, or remove it, or replace with if (...) { log_err(...) } anyway?_**

Few lines above there is another assert: assert(src_info); Which is also always true unless someone introduce a bug. But for this assert() I agree it is better to remove it as in the worst case src_info == NULL we will get a proper fatal handling when trying to access src_info and no magic issues.

paulstelian97 replied:
Don't use asserts alone as an error, because the compiler is allowed to remove them. If it's important that the firmware errors and recovers (which it probably is) then proper error handling should be added in addition to the assert. If the assert can be triggered by invalid inputs (rather than _just_ programming errors) it's invalid as an assert.

paulstelian97 replied:
Still, what should be done if the assert is gone? Do something else?

lgirdwood reviewed (commented):
In general this is almost ready to merge. We do need to 
1) remove/reduce the usage of assert() and use log() and send the error up the stack instead.
2) Some large code blocks need comments decribing what we are doing.
3) Minor - C style comments.

serhiy-katsyuba-intel commented:
Made some minor style changes:
* replaced nice // comments to ugly /* */
* added more explanatory comments
* removed few unnecessary asserts
* rebased.

lgirdwood reviewed (approved):
@lyakh fyi - you may need to manually update for SPARSE when back as we can't yet enforce SPARSE correctness until enabled in CI and all PRs merged..

serhiy-katsyuba-intel commented:
> @lyakh fyi - you may need to manually update for SPARSE when back as we can't yet enforce SPARSE correctness until enabled in CI and all PRs merged..

I used __sparse_cache in mixin_mixout.c where necessary, so (if I understand the message correctly :) ) no need for @lyakh to do anything with __sparse_cache there.

lgirdwood commented:
> 

Great - my comment was more to make @lyakh aware that we have a new files that should be included for SPARSE correctness. Fwiw, there are still a few SPARSE PRs to merge before we can fully run the correctness checking.
